src/operators/stateful_unary.rs

```diff
@@ -412,6 +401,12 @@ where
                             output_session
                                 .give_iterator(output.into_iter().map(|item| (key.clone(), item)));
 
+                            if let Some(next_awake) = logic.next_awake() {
+                                current_next_awake.insert(key.clone(), next_awake);
+                            } else {
+                                current_next_awake.remove(&key);
+                            }
```

Comment:
This is what I was referring to in the last paragraph of the PR description. We need to note when the logic for each key should be awoken next irrespective of the epoch interval. If a system time window is going to close in 3 sec, we need to mark that ASAP, not just when the epoch closes (which is where this block was before).

That is a good point, though, that we also probably should be re-calculating fate after each logic call as well, and not just at the end of the epoch. I believe as is is "correct" as long as each logic returns to a "zero state" to signal ready for discard, since in that case if the same key comes back up within the same epoch, it'll be played into an effectively-new logic. I think all our stateful operators have pretty simple ways to determine fate, so I don't think they trigger this. I'll refactor this to call fate continuously, though.